### ISSUE TITLE
Question: Does c2rust compatible with LLVM 9?

### ISSUE THREAD

**Issue by simnalamburt** [immunant/c2rust#245]
I'm trying to use c2rust to transpile Linux kernel modules into Rust. c2rust uses LLVM backends which do not support GNU C Extensions so transpiling Linux kernel modules with c2rust is tricky. However, with LLVM 9, it became much easier since many improvements regarding C Extensions landed in LLVM 9.

> C2Rust requires LLVM 6, 7, or 8 with its corresponding clang compiler and libraries.

The README says that it requires LLVM 6, 7, or 8 but looks like some works to support LLVM 9 is going on: https://github.com/immunant/c2rust/commit/3bacb538. Does c2rust 0.14.0 or c2rust `master` branch supports LLVM 9?

**Comment by ahomescu** [comment]
The main GNU extension that the kernel uses from LLVM 9 are _asm gotos_. We should be able to add general support for LLVM 9 to C2Rust, but asm gotos are a special case. AFAIK there is no equivalent Rust syntax for that particular inline assembly feature, so C2Rust wouldn't be able to transpile them asm gotos at all.

> Does c2rust 0.14.0 or c2rust master branch supports LLVM 9?

Not explicitly, because we know that asm gotos aren't supported and we haven't tested anything else. It might work if you try it on your own kernel modules.

**Comment by simnalamburt** [comment]
Yes that's right. asm gotos are the main blocker of using c2rust with Linux kernel modules and asm goto support has landed on LLVM 9. I tried it and got "AST node not found", looks like LLVM succeeded with transpiling Linux headers but c2rust failed to understand it.

Then I have another question. Looks like c2rust transpiles all codes of headers even if they're not required to transpile. Is there any good way to skip transpile of functions in headers? Those asm goto syntaxes are only used in headers in most projects.

**Comment by thedataking** [comment]
@simnalamburt can you let us know what kernel module you are compiling and the file your screenshot is from?

Regarding your question about unused functions in headers. We have a heuristic that removes header declarations that are unused *and* not exported. If an item is externally visible, we have to keep it to remain faithful to the semantics of C. If you can show us a header declaration that is both unused and not externally visible, we would appreciate it if you can share it.

**Comment by simnalamburt** [comment]
@thedataking It was https://github.com/simnalamburt/flashbench.
